ably-ios/ARTEventEmitter.m

```diff
@@ -6,39 +6,266 @@
+    [self onAllCallListener:listener];
+    return listener;
+}
+
+- (void)onAllCallListener:(ARTEventListener *)listener {
+    [self addOnAllEntry:[[ARTEventEmitterEntry alloc] initWithListener:listener once:false]];
+}
+
+- (ARTEventListener *)onceAll:(void (^)(id __art_nonnull))cb {
+    ARTEventListener *listener = [[ARTEventListener alloc] initWithBlock:cb];
+    [self onceAllCallListener:listener];
+    return listener;
+}
+
+- (void)onceAllCallListener:(ARTEventListener *)listener {
+    [self addOnAllEntry:[[ARTEventEmitterEntry alloc] initWithListener:listener once:true]];
+}
+
+- (void)addOnAllEntry:(ARTEventEmitterEntry *)entry {
+    [self.totalListeners addObjectReplacing:entry];
+    
+    // Maybe cb was already listening to some events; remove it from their
+    // entries in listeners if so.
+    for (NSMutableArray *listenersToEvent in [self.listeners allValues]) {
+        [listenersToEvent artRemoveObject:entry.listener];
+    }
+}
+
+- (void)off:(id)event listener:(ARTEventListener *)listener {
+    if ([self.totalListeners member:listener] != nil) {
+        // It is listening to everything but now wants to ignore a particular
+        // event. Just mark it in its ignoring entry; emit will check that.
+        [self addObject:listener toArrayWithKey:event inDictionary:self.ignoring];
+        return;
+    }
+
+    [self removeFromArray:[self.listeners objectForKey:event] item:listener];
+}
+
+- (void)offAll:(ARTEventListener *)listener {
+    [self.totalListeners artRemoveObject:listener];
+    for (NSMutableArray *listeners in [self.ignoring allValues]) {
+        [listeners artRemoveObject:listener];
+    }
+    for (NSMutableArray *listenersToEvent in [self.listeners allValues]) {
+        [listenersToEvent artRemoveObject:listener];
+    }
+}
+
+- (void)emit:(id)event with:(id)data {
+    NSMutableArray *listenersForEvent = [self.listeners objectForKey:event];
+    NSMutableArray *toCall = [[NSMutableArray alloc] init];
+    NSMutableArray *toRemoveFromListenersForEvent = [[NSMutableArray alloc] init];
+    NSMutableArray *toRemoveFromTotalListeners = [[NSMutableArray alloc] init];
+    @try {
+        for (ARTEventEmitterEntry *entry in listenersForEvent) {
+            if (entry.once) {
+                NSMutableArray *ign = [self.ignoring objectForKey:event];
+                if (ign && [ign member:entry.listener]) {
+                    // If you call 'onAll', then 'once(A)' and then 'off(A)',
+                    // 'onAll' will add to totalListeners, 'once(A)' will add
+                    // to listeners and 'off(A)' will add to ignoring. We've found
+                    // here one of those cases; just keep going. (If 'on' or 'once'
+                    // is called again, it will remove A from ignoring and replace this
+                    // entry, so it's fine to leave it here.)
+                    [toRemoveFromListenersForEvent addObjectReplacing:entry];
+                    continue;
```

Comment:
I see why that's nice, but I think it's too complicated from a developer's perspective to keep track of what their ignore list is.  Lets discuss today with @paddybyers though as we need to agree on a simple spec so that you don't waste time on this, sorry the spec was not clear enough in the first place.